src/Compilers/Core/Portable/Operations/IOperation.ChildCollection.cs

```diff
@@ -0,0 +1,142 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Collections;
+using System;
+using System.Diagnostics;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis
+{
+    public partial interface IOperation
+    {
+        /// <summary>
+        /// Implements a struct-based collection of <see cref="Operation"/> nodes. This collection is ordered, but
+        /// random access into the collection is not provided. This type is not hardened to 
+        /// <code>default(ChildCollection)</code>, and will null reference in these cases.
+        /// </summary>
+        [NonDefaultable]
+        public readonly partial struct ChildCollection : IReadOnlyCollection<IOperation>
+        {
+            private readonly Operation _operation;
+
+            internal ChildCollection(Operation operation)
+            {
+                _operation = operation;
+            }
+
+            public int Count => _operation.ChildOperationsCount;
+
+            public Enumerator GetEnumerator() => new Enumerator(_operation);
+
+            public ImmutableArray<IOperation> ToImmutableArray()
+            {
+                Enumerator enumerator = GetEnumerator();
+                switch (_operation)
+                {
+                    case NoneOperation { Children: var children }:
+                        return children;
+                    case InvalidOperation { Children: var children }:
+                        return children;
+                    case var _ when !enumerator.MoveNext():
+                        return ImmutableArray<IOperation>.Empty;
+                    default:
+                        var builder = ArrayBuilder<IOperation>.GetInstance(Count);
+                        do
+                        {
+                            builder.Add(enumerator.Current);
+                        } while (enumerator.MoveNext());
+
+                        return builder.ToImmutableAndFree();
+                }
+            }
+
+            IEnumerator<IOperation> IEnumerable<IOperation>.GetEnumerator() => this.GetEnumerator();
+            IEnumerator IEnumerable.GetEnumerator() => this.GetEnumerator();
+
+            public bool Any() => Count > 0;
+
+            public IOperation First()
+            {
+                var enumerator = GetEnumerator();
+                if (enumerator.MoveNext())
+                {
+                    return enumerator.Current;
+                }
+
+                throw new InvalidOperationException();
+            }
+
+            public Reversed Reverse() => new Reversed(_operation);
```

Comment:
>Reverse() [](http://example.com/codeflow?start=28&length=9)

Where is `Reverse()` used?

Reply:
API review asked for the APIs we expose for the syntax version of this list to be added, including `Reverse()`. In this particular PR, we use it for implementing `Last()`, without having to walk the entire collection.

Reply:
Is there a test for `Reverse()`? If we're testing the forward iterators of each `IOperation` implementation, please consider testing that `Reverse()` returns the reverse order.